pkg/proxy/ipvs/graceful_termination.go

```diff
@@ -82,8 +82,12 @@ func (q *graceTerminateRSList) remove(rs *listItem) bool {
 }
 
 func (q *graceTerminateRSList) flushList(handler func(rsToDelete *listItem) (bool, error)) bool {
+	q.lock.Lock()
+	list := q.list
+	q.lock.Unlock()
+	
 	success := true
-	for name, rs := range q.list {
+	for name, rs := range list {
```

Comment:
Last time I checked, maps are passed by reference so you are still reading q.list without a lock here. Does your static check say differently?

Reply:
Oh, my bad I didn't look at the code since the last push. Before it just added the lock and still used q.list. @andrewsykim you are correct, this would do nothing.

Reply:
Either that or i am remembering wrong...

Reply:
Oh, you are remembering perfectly correct. It is my fault. In the previous commit I didn't use `list := q.list`, and that commit will cause a double lock bug. I fixed it later.

Reply:
ah. well, then this one doesn't do anything...which is why there is no double locking

Reply:
OMG. Actually with this commit I changed a data race bug into an atomicity violation bug... 
I can think of a kinda weird patch: 
```Go
func (q *graceTerminateRSList) remove(rs *listItem, caller_is_holding_Lock bool) bool {
	if caller_is_holding_Lock {
		// do nothing
	} else {
		q.lock.Lock()
		defer q.lock.Unlock()
	}
	...
}
```

Reply:
At first glance it makes the code look a bit awkward. Can we list the entries in a separate method and pass it into `flushList?

Reply:
Where is the double-lock happening? That is, what is flushList calling that also tries to acquire the lock?

Reply:
`q.remove(rs)` on line 98 uses the lock

Reply:
Currently there are 2 places using `remove`. One is at line 98, which holds the lock. One is at line 205, which doesn't hold the lock. 
@andrewsykim I am not sure if I got your idea correctly. Here is my understanding: we will have 2 methods: `remove` and `removeWithLock`, and let line 98 calls the former, let line 205 calls the later.
BTW, please tell me if there is a better name than `removeWithLock`. I am really bad at this~
```Go
func (q *graceTerminateRSList) removeWithLock(rs *listItem) bool {
	q.lock.Lock()
	defer q.lock.Unlock()

	return q.remove(rs)
}

func (q *graceTerminateRSList) remove(rs *listItem) bool {
	uniqueRS := rs.String()
	if _, ok := q.list[uniqueRS]; ok {
		delete(q.list, uniqueRS)
		return true
	}
	return false
}
```